plugins/cache-caffeine/src/test/java/CaffeineHeapCacheTests.java

```diff
@@ -0,0 +1,451 @@
+                    public boolean isLoaded() {
+                        return isLoaded;
+                    }
+
+                    @Override
+                    public String load(ICacheKey<String> key) {
+                        isLoaded = true;
+                        return value;
+                    }
+                };
+                loadAwareCacheLoaderList.add(loadAwareCacheLoader);
+                phaser.arriveAndAwaitAdvance();
+                try {
+                    assertEquals(value, caffeineTest.computeIfAbsent(getICacheKey(key), loadAwareCacheLoader));
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+                countDownLatch.countDown();
+            });
+            threads[i].start();
+        }
+        phaser.arriveAndAwaitAdvance();
+        countDownLatch.await();
+        int numberOfTimesValueLoaded = 0;
+        for (int i = 0; i < numberOfRequest; i++) {
+            if (loadAwareCacheLoaderList.get(i).isLoaded()) {
+                numberOfTimesValueLoaded++;
+            }
+        }
+        assertEquals(1, numberOfTimesValueLoaded);
+        assertEquals(1, caffeineTest.stats().getTotalMisses());
+        assertEquals(1, caffeineTest.stats().getTotalItems());
+        assertEquals(numberOfRequest - 1, caffeineTest.stats().getTotalHits());
+        assertEquals(1, caffeineTest.count());
+        caffeineTest.close();
+    }
+
+    public void testInvalidateAll() throws Exception {
+        ToLongBiFunction<ICacheKey<String>, String> weigher = getMockWeigher();
+        MockRemovalListener<String, String> removalListener = new MockRemovalListener<>();
+        ICache<String, String> caffeineTest = new CaffeineHeapCache.Builder<String, String>()
+            .setDimensionNames(List.of(dimensionName))
+            .setExecutor(Runnable::run)
+            .setExpireAfterAccess(TimeValue.MAX_VALUE)
+            .setMaximumWeightInBytes(CACHE_SIZE_IN_BYTES)
+            .setWeigher(weigher)
+            .setRemovalListener(removalListener)
+            .build();
+        int randomKeys = randomIntBetween(10, 100);
+        Map<ICacheKey<String>, String> keyValueMap = new HashMap<>();
+        for (int i = 0; i < randomKeys; i++) {
+            keyValueMap.put(getICacheKey(UUID.randomUUID().toString()), UUID.randomUUID().toString());
+        }
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            caffeineTest.put(entry.getKey(), entry.getValue());
+        }
+        caffeineTest.invalidateAll(); // clear all the entries.
+        for (Map.Entry<ICacheKey<String>, String> entry : keyValueMap.entrySet()) {
+            // Verify that value is null for a removed entry.
+            assertNull(caffeineTest.get(entry.getKey()));
+        }
+        assertEquals(0, caffeineTest.count());
+        caffeineTest.close();
+    }
+
+    public void testInvalidate() throws Exception {
```

Comment:
We can probably remove this test, and only use testInvalidateConcurrently, as it's kind of a superset of this test's behavior